### ISSUE TITLE
CSV.read not working properly with StructArray

### ISSUE THREAD

**Issue by robsmith11** [JuliaData/CSV.jl#941]
It works as expected when reading into a `NamedTulpe` first, but not when reading directly into a `StructArray`:

```
julia> write("/tmp/a.csv", "a,b,c\n1,2,3\n4,5,6\n");

julia> CSV.read("/tmp/a.csv", StructArrays.StructArray)
0-element StructArray() with eltype Vector{Int64} with indices 1:0

julia> StructArrays.StructArray(CSV.read("/tmp/a.csv", NamedTuple))
2-element StructArray(::Vector{Int64}, ::Vector{Int64}, ::Vector{Int64}) with eltype NamedTuple{(:a, :b, :c), Tuple{Int64, Int64, Int64}}:
 (a = 1, b = 2, c = 3)
 (a = 4, b = 5, c = 6)

julia> Pkg.status("CSV")
Status `/me/.julia/environments/v1.8/Project.toml`
  [336ed68f] CSV v0.9.10
```

**Comment by quinnj** [comment]
Pinging @piever to chime in, but as I understand it, a `StructArray` is a valid Tables.jl _source_, but doesn't currently have a _sink_ interface. i.e. the generic fallback `StructArray(x)` definitions expects `x` to be a struct iterator, but not necessarily a Tables.jl-related source.

My guess is it would be trivial to add some kind of `StructArrays.fromtable` function that would take any valid Tables.jl source and create a StructArray, if you wanted to take a stab at a PR.

**Comment by piever** [comment]
Yes, that is correct. You can actually already do

```julia
CSV.read(fn, StructArray∘Tables.columntable),
```

which is probably close to optimal (the data structure backing a `StructArray` is a named tuple of vectors).

Should `CSV.read` call `materializer(sink)` rather than `sink` to process the data? That way, I can just define the correct

```julia
materializer(::Type{<:StructArray})`
```

method, and `CSV.read(fn, StructArray)` should just work.

**Comment by quinnj** [comment]
It seems like `StructArray∘Tables.columntable` is the way to go to me. `CSV.read` is pretty well documented that you need to pass a valid sink function; `materializer` is usually just for when you pass in an _instance_ of a table and want to get its materializing function.

**Comment by robsmith11** [comment]
I think most users who aren't familiar with the implementation details are still going to expect StructArray to work as a sink without going through an intermediate step.   It's a table-like data structure just like DataFrame.

**Comment by piever** [comment]
> `CSV.read` is pretty well documented that you need to pass a valid sink function; 

Makes sense. I can probably still define `materializer` for a `StructArray` type, so that users can do

```julia
CSV.read(fn, materializer(StructArray))
```

Rather than defining `CSV.read(fn, sink::Type)` as `CSV.File(fn) |> materializer(sink)` (it may be a step too far) maybe the docs of `CSV.read` could briefly mention the `materializer` helper to get a valid sink function from an instance or type (I think both work) of table.

**Comment by quinnj** [comment]
Direct StructArray sinks now work with CSV.jl 1.0 and StructArrays 0.7.3 ([#1196](https://github.com/JuliaData/CSV.jl/pull/1196)). [issue resolution reviewed by AI; please respond with objections or additional feedback/concerns/issues]
